flagext/deprecated.go

```diff
@@ -16,22 +17,170 @@ var DeprecatedFlagsUsed = promauto.NewCounter(
 		Help: "The number of deprecated flags currently set.",
 	})
 
-type deprecatedFlag struct {
-	name   string
-	logger log.Logger
+// DisabledFlagsUsed is the metric that counts deprecated flags set.
+var DisabledFlagsUsed = promauto.NewCounter(
+	prometheus.CounterOpts{
+		Name: "disabled_flags_inuse_total",
+		Help: "The number of disabled flags currently set.",
+	})
+
+type disabledFlag struct {
+	name              string
+	countAsDeprecated bool
+	logger            log.Logger
 }
 
-func (deprecatedFlag) String() string {
+func (disabledFlag) String() string {
 	return "deprecated"
 }
 
-func (d deprecatedFlag) Set(string) error {
+func (d disabledFlag) Set(string) error {
 	level.Warn(d.logger).Log("msg", "flag disabled", "flag", d.name)
-	DeprecatedFlagsUsed.Inc()
+	if d.countAsDeprecated {
+		DeprecatedFlagsUsed.Inc()
+	} else {
+		DisabledFlagsUsed.Inc()
+	}
 	return nil
 }
 
-// DeprecatedFlag logs a warning when you try to use it.
+// DeprecatedFlag registers a noop flag and logs a warning when you try to use it and increments the DeprecatedFlagsUsed counter.
+//
+// This function is deprecated and will be removed in the future. Please use DisabledFlag or the DeprecatedVar alternatives.
 func DeprecatedFlag(f *flag.FlagSet, name, message string, logger log.Logger) {
-	f.Var(deprecatedFlag{name: name, logger: logger}, name, message)
+	f.Var(disabledFlag{name: name, logger: logger, countAsDeprecated: true}, name, message)
+}
+
+// DisabledFlag registers a noop flag and logs a warning when you try to use it and increments the DisabledFlagsUsed counter.
+func DisabledFlag(f *flag.FlagSet, name, message string, logger log.Logger) {
+	f.Var(disabledFlag{name: name, logger: logger, countAsDeprecated: false}, name, message)
+}
+
+// deprecatedFlag wraps a flag.FlagSet so we can add behaviour on Set without having to reimplement the parsing.
+type deprecatedFlag struct {
+	name     string
+	logger   log.Logger
+	delegate *flag.FlagSet
+}
+
+func (d deprecatedFlag) String() (s string) {
+	// The delegate should only have a single registered flag.
+	d.delegate.VisitAll(func(f *flag.Flag) {
+		s = f.Value.String()
+	})
+	return
+}
+
+func (d deprecatedFlag) Set(s string) error {
+	level.Warn(d.logger).Log("msg", "flag deprecated", "flag", d.name)
```

Comment:
I'm not 💯 sure. In which case are we going to use it? I think we should also warn users when using YAML (most of OSS community use YAML, not CLI flags), so I can't see in which case we're going to use these new flags.

Reply:
i was more thinking about dskit on its own, not how we'll use it in Mimir in particular. Currently the `DeprecatedFlag` function is not doing what we really want it to do.

However, I'm not that keen to merge this since it's not solving any problem I/we have anymore. Should I leave it open to get more feedback or close it right away?

Reply:
Sure, let's hear some feedback from others too.